AGENTS.md

```diff
@@ -0,0 +1,165 @@
+
+- ✅ **DO**: Use `@Provides` methods in modules to create instances
+  ```java
+  @Provides
+  @Singleton
+  public MyService myService() {
+      return new MyServiceImpl();
+  }
+  ```
+- ✅ **DO**: Use qualified injections with `@Named` when multiple instances of the same type exist
+  ```java
+  @Provides
+  @Singleton
+  @Named("java-flight-recorder")
+  public ObjectMapper objectMapper() {
+      return new ObjectMapper();
+  }
+  ```
+- ❌ **DON'T**: Use `@Named` directly on implementation classes without a corresponding `@Provides` method
+  - **Why**: Violates Guice best practices; prefer explicit provider methods for clarity
+
+### Java Language Features
+
+- ✅ **DO**: Use Java records for simple data classes
+  ```java
+  public record JfrSession(
+      @JsonProperty @NonNull String name,
+      @JsonProperty long id) {}
+  ```
+- ✅ **DO**: Use JSpecify annotations (`@NonNull`, `@Nullable`) for nullability contracts
+- ✅ **DO**: Use modern Java features (streams, lambdas, etc.) - this project targets Java 17+
+
+### Internationalization
+
+- ✅ **DO**: Use full, descriptive names in Messages.properties
+  - Example: `JfrRootAction.DisplayName=Java Flight Recorder` (not "JFR")
+- ✅ **DO**: Follow the pattern `ClassName.PropertyName=Value`
+
+## Implementation Guidelines
+
+### Using JFR APIs
+
+When working with Java Flight Recorder:
+
+- ✅ **DO**: Use `jdk.jfr.FlightRecorder` and related classes to get real runtime data
+  ```java
+  FlightRecorder.getFlightRecorder().getRecordings()
+  ```
+- ❌ **DON'T**: Use hardcoded or mock data in production code
+
+### JSON Serialization
+
+- ✅ **DO**: Use Jackson's `ObjectMapper` for JSON serialization
+- ✅ **DO**: Inject the `ObjectMapper` with proper qualifiers
+- ✅ **DO**: Use `@JsonProperty` annotations on record components for explicit serialization
+
+## File Organization
+
+### Maven Settings
+
+- ❌ **NEVER** commit Maven settings files to the repository
+  - Files like `~/.m2/settings.xml` belong in the user's home directory, not in source control
+  - The `~` directory should never appear in the repository
+
+### Source Structure
+
+Follow standard Maven project structure:
+```
+src/
+├── main/
+│   ├── java/
+│   │   └── io/jenkins/plugins/jfr/
+│   └── resources/
+│       └── io/jenkins/plugins/jfr/
+└── test/
+    └── java/
+        └── io/jenkins/plugins/jfr/
+```
+
+## Testing
+
+- ✅ **DO**: Write unit tests for new functionality
+- ✅ **DO**: Use JUnit 5 (Jupiter) for tests
+- ✅ **DO**: Use Mockito for mocking when appropriate
+- ✅ **DO**: Follow the given-when-then pattern in tests
+  ```java
+  @Test
+  public void testGetSessions() {
+      // given
+      Collection<JfrSession> sessions = Arrays.asList(new JfrSession("test", 123));
+      given(service.getSessions()).willReturn(sessions);
+      
+      // when
+      Collection<JfrSession> result = action.getSessions();
+      
+      // then
+      assertEquals(sessions, result);
```

Comment:
```suggestion
      assertThat(sessions).isEqualTo(result);
```

Favor assertj assertions for better readability and quicker time to diagnosis when dealing with failed tests.